File: 剑指Offer24.反转链表/剑指Offer24-反转链表.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 按照饲养员解法超出时间限制
        # dummy=ListNode(0)
        # dummy.next=head
        # while head is not None and head.next is not None:
        #     dnext=dummy.next
        #     hnext=head.next
        #     dummy.next=hnext
        #     head.next==hnext.next
        #     hnext.next=dnext
        # return dummy.next
        
        #参考解法
        pre=None
        cur=head
        while cur is not None:
            temp=cur.next
            cur.next=pre
            # 更新迭代
            pre=cur
            cur=temp
        return pre

File: 剑指Offer24.反转链表/剑指Offer24-反转链表.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* reverseList(struct ListNode* head){
    struct ListNode* pre = (struct ListNode*)malloc(sizeof(struct ListNode));
    pre = NULL;
    struct ListNode* cur = (struct ListNode*)malloc(sizeof(struct ListNode));
    cur = head;
    while(cur){
        struct ListNode* tmp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = tmp;
    }
    return pre;

}